Extensible validation support in Volto forms (#6181)
plone/volto: docs/source/configuration/validation.md

---
myst:
  html_meta:
    "description": "Client side form field validation"
    "property=og:description": "Client side form field validation"

>     "property=og:title": "Form fields validation"

Reviewer: ```suggestion
    "property=og:title": "Client side form field validation"
```

      description: '',
      format: 'url'
    },
  },
  required: [],
};
```

`url` named validator should be registered by this name as a Volto validator utility:

```ts
config.registerUtility({
  type: 'validator',
  name: 'url',
  method: urlValidator,
})
```

In this case, the `urlValidator` method validator will be applied for the block field `customField`.

#### For content types

Content types also can specify the `format` using the schema hints in the backend in the `frontendOptions`:

```python
from plone.supermodel import model
from zope import schema

class IMyContent(model.Schema):
    directives.widget(
        "customField",
        frontendOptions={
            "format": "url",
        },
    )
    customField = schema.TextLine(
        title="Custom URL field",
        required=False,
    )
    # Rest of your content type definition
```

For the record, the resultant `plone.restapi` response will be something like the following, a bit different than in blocks JSON schema.
But the validation engine will take care too:

```json
{
  "properties": {
    "customField": {
      "title": "Custom URL field",
      "widgetOptions": {
        "frontendOptions": {

>           "format": "url"

Reviewer 2: Can we change the api call so it will be "flattened" to the root of the field so we only have 1 way specifying the format?

Author: I had this wish back in the day too, but it never happened, since we already had this "scape hatch" for all customizable things in Python schema hints. Maybe we can push again in the backend side for this to happen.

@davisagli do you think we could try to unify this also in the backend? Or do you remember the reason behind we never did it?

Reviewer 3: It depends on the property. For things that are valid JSON Schema, it would make sense to have plone.restapi serialize them at the top level. For other properties that are not part of JSON Schema, it makes sense to keep them inside the widgetOptions.

        }
      }
    }
  }
}
```

and the `urlValidator` method validator will be applied for the content type field `customField`.

### Advanced scenarios

In case you need more granularity and you don't have access to modify the existing implementation of the JSON schema definitions for existing content types, blocks or forms (might be in third party add-ons), you can use the following advanced validator registrations, using `field`, `widget`, `behaviorName` or `blockType` validator registrations.

#### Per field `type` validators

These validators are applied depending on the specified `type` of the field in the JSON schema from content types, forms or blocks.
The next example is for the use case of JSON schema defined in a block:

```ts
let blockSchema = {
  // ... fieldset definition in here
  properties: {
    ...schema.properties,
    customField: {
      title: 'My custom field',
      description: '',
      type: 'integer',
      maximum: 30
    },
  },
  required: [],
};
```

```ts
config.registerUtility({
  type: 'validator',

>   name: 'maximum',

Reviewer 2: This is indeed part of the validation spec of json schema (same for maxLength, minLength, maxItems, pattern etc). Should we make these validators customisable? I think it would make more sense to have these "build-in" as validators since the name already implies what it should do.

Author: What we have today as built in:
- minLength
- maxLenght
- minimum
- maximum
- uniqueItem

We could implement the rest, for sure.

  dependencies: {
    fieldType: 'integer',
  },
  method: maximumValidator,
})
```

You should specify the `type` in the JSON schema of the block (in a content type, it is included in the default serialization of the field).
If a field does not specify type, it assumes a `string` type as validator.

#### Per field `widget` validators

> These validators are applied depending on the specified `widget` of the field.

Reviewer 2: In the above examples you have to specify in the json schema if the validator is enabled or not. In the registerUtility method you only specify if the validator is available or not. It seems in this example registering the validator will have it enabled by default. I think for consistency we should also specify it in the schema and not enable it by default.

Author: I will clarify which validators are included and enabled by default, and put it on the top, instead than at the bottom.

Reviewer: Yeah, having that at the end of the docs, with a `seealso` where it was first mentioned, didn't feel right. This would be a good change.

Author: @stevepiercy moved to the top.

---
myst:
  html_meta:
    "description": "Client side form field validation"
    "property=og:description": "Client side form field validation"
    "property=og:title": "Client side form field validation"
    "keywords": "Volto, Plone, frontend, React, configuration, form, fields, validation"
---

# Client side form field validation

> Volto provides a mechanism for delivering form field validation in an extensible way.

Reviewer 3: ```suggestion
Volto provides an extensible way to validate form field values.
```

```ts
config.registerUtility({
  type: 'validator',
  name: 'url',
  method: urlValidator,
})
```

#### Content types

You can also specify the `format` of content types using the schema hints in the backend using `frontendOptions`.

```python
from plone.supermodel import model
from zope import schema

class IMyContent(model.Schema):
    directives.widget(
        "customField",
        frontendOptions={
            "format": "url",
        },
    )
    customField = schema.TextLine(
        title="Custom URL field",
        required=False,
    )
    # Rest of your content type definition
```

The response from `plone.restapi` will be something like the following.
It is slightly different from blocks JSON schema, but the validation engine will behave the same.
The `urlValidator` method validator will be applied for the content type field `customField` from the earlier example.

```json
{
  "properties": {
    "customField": {
      "title": "Custom URL field",
      "widgetOptions": {
        "frontendOptions": {
          "format": "url"
        }
      }
    }
  }
}
```

### Advanced scenarios

> In case you need more granularity, and you don't have access to modify the existing implementation of the JSON schema definitions for existing content types, blocks, or forms, you can use the following advanced validator registrations, using `field`, `widget`, `behaviorName`, or `blockType` validator registrations.

Reviewer 3: I don't think "more granularity" is the right way to describe this. Nothing is more granular (specific) than assigning one specific validator to a specific field using `format`. This is less granular: it's a generic approach to register validators that apply to multiple fields automatically based on their field/widget/behaviorName/blockType. Right?

Author: @davisagli I tried to improve it, take a look if now it's better.

The next example shows how to define the JSON schema in a block.

```ts
let blockSchema = {
  // ... fieldset definition in here
  properties: {
    ...schema.properties,
    customField: {
      title: 'My custom field',
      description: '',
      type: 'integer',
      maximum: 30
    },
  },
  required: [],
};
```

```ts
config.registerUtility({
  type: 'validator',
  name: 'maximum',
  dependencies: {
    fieldType: 'integer',
  },
  method: maximumValidator,
})
```

#### Field `widget` validators

Field `widget` validators are applied depending on the specified `widget` of the field.
You should specify the `widget` either in the JSON schema of the block or as additional data in the content type definition.

The following example shows how to specify the `widget` either in the JSON schema of the block.

```ts
let blockSchema = {
  // ... fieldset definition in here
  properties: {
    ...schema.properties,
    customField: {
      title: 'My custom field',
      description: '',
      widget: 'phoneNumber',
    },
  },
  required: [],
};
```

```ts
config.registerUtility({
  type: 'validator',
  name: 'phoneNumber',
  dependencies: {

>     widgetName: 'phoneNumber',

Reviewer 3: Why is it `widget` in the schema but `widgetName` here?

Author: Moved to use `widget` only.

  },
  method: phoneValidator,
})
```

The following example shows how to specify the `widget` in the content type definition in the schema hints in the backend using `frontendOptions`.
The validation engine  will behave the same, applying the `urlValidator` method validator for the content type field `customField` in the previous example.

```python
from plone.supermodel import model
from zope import schema

class IMyContent(model.Schema):
    directives.widget(
        "customField",
        frontendOptions={
            "widget": "url",
        },
    )
    customField = schema.TextLine(
        title="Custom URL field",
        required=False,
    )
    # Rest of your content type definition
```

#### Behavior and field name validators

Behavior and field name validators are applied depending on the behavior in combination with the name of the field.
These usually come from a content type definition.
This type of validator only applies to content type validators.
It takes the `behaviorName` and the `fieldName` as dependencies.

```ts
config.registerUtility({
  type: 'validator',
  name: 'dateRange',
  dependencies: {

>     behaviorName: 'plone.eventbasic',

Reviewer 3: @stevepiercy This behavior has been defined in plone.app.event for years. It can't be changed without breaking backwards-compatibility.

    fieldName: 'start'
  },
  method: startEventDateRangeValidator,
})
```

#### Block type and field name validators

Block type and field name validators are applied depending on the block type in combination with the name of the field in the block settings JSON schema.
This type of validator only applies to blocks.
It takes the `blockType` and the `fieldName` as dependencies.

```ts
config.registerUtility({
  type: 'validator',
  name: 'url',
  dependencies: {
    blockType: 'slider',
    fieldName: 'url'
  },
  method: urlValidator,
})
```

(voltos-default-validators-label)=

## Volto's default validators

Volto provides a set of validators by default.

> You can find them in the module {file}`packages/volto/src/config/validators.ts`.

Reviewer 3: We should ideally document them here rather than just pointing to source code.

Author: Done.

# Client side form field validation

Volto provides an extensible way to validate form field values.
This extensibility is based on the Volto registry.
It applies to content types, custom programatically generated forms, and blocks schema settings.
The mechanism serializes all of them according to the [JSON schema standard](https://json-schema.org/draft/2020-12/json-schema-validation).
Finally Volto generates the form from the serialization.

(voltos-default-validators-label)=

## Volto's default validators

Volto provides a set of validators by default:

### Strings
- minLength
- maxLength
- pattern

### Password
- minLength
- maxLength
- pattern

### Numbers
- isNumber
- minimum
- maximum

### Integers
- isInteger
- minimum
- maximum

### Arrays
- maxItems
- minItems
- uniqueItems

### Per widget
- email
- url

### Event content type

> - start/end dates check

Reviewer: ```suggestion
- event end date must be on or after its start date
```

### Advanced scenarios

> If, for some reason, you can't modify the existing implementation of the JSON schema definitions for existing content types, blocks, or forms, you can use the following advanced validator registrations, which allows you to register validators depending on the `field`, the `widget`, the `behaviorName` (in case of a content type), or the `blockType` (in case of a block).

Reviewer: ```suggestion
If, for some reason, you can't modify the existing implementation of the JSON schema definitions for existing content types, blocks, or forms, you can use the following advanced validator registrations.
These allow you to register validators according to whether it is a `field`, `widget`, `behaviorName` (for content types), or `blockType` (for blocks).
```

Using `formData`, you can perform validation checks using other field data as source.
This is useful when you want to validate two related fields, such as ensuring the end date of an event is after its start date.
You can create invariant validator types.

> In the following code snippet you can check how to create a validator method for checking if the Event content type `start`/`end` fields are valid:

Reviewer: Note that this narrative text reflects that the start and end dates of an event content type are different. I think I read this code sample correctly, but another pair of non-jetlagged eyes would be good.

```suggestion
The following code snippet shows how to create a validator method that ensures the event content type's end date is after its start date.
```
